apps/els_lsp/src/els_bsp_provider.erl

```diff
@@ -0,0 +1,147 @@
+            , <<"version">>      => list_to_binary(Vsn)
+            , <<"bspVersion">>   => <<"2.0.0">>
+            , <<"rootUri">>      => RootUri
+            , <<"capabilities">> => #{ <<"languageIds">> => [<<"erlang">>] }
+            , <<"data">>         => #{}
+            },
+  request(<<"build/initialize">>, Params, State);
+handle_info({request_finished, Request, Response}, State) ->
+  handle_response(Request, Response, State);
+handle_info(Msg, State) ->
+  case check_response(Msg, State) of
+    {ok, NewState} ->
+      NewState;
+    no_reply ->
+      ?LOG_WARNING("Discarding unrecognized message: ~p", [Msg]),
+      State
+  end.
+
+%%==============================================================================
+%% Internal functions
+%%==============================================================================
+-spec request(binary(), map() | null, state()) -> state().
+request(Method, Params, #{ pending := Pending } = State) ->
+  RequestId = els_bsp_client:request(Method, Params),
+  State#{ pending => [{RequestId, {Method, Params}} | Pending] }.
+
+-spec handle_response({binary(), any()}, any(), state()) -> state().
+handle_response({<<"build/initialize">>, _Params}, Response, State) ->
+  ?LOG_INFO("BSP Server initialized: ~p", [Response]),
+  ok = els_bsp_client:notification(<<"build/initialized">>),
+  request(<<"workspace/buildTargets">>, #{}, State);
+handle_response({<<"workspace/buildTargets">>, _Params}, Response, State0) ->
+  Result = maps:get(result, Response, #{}),
+  Targets = maps:get(targets, Result, []),
+  TargetIds = lists:flatten([ maps:get(id, Target, []) || Target <- Targets ]),
+  State1 = request(<<"buildTarget/sources">>, #{ <<"targets">> => TargetIds }, State0),
+  State2 = request(<<"buildTarget/dependencySources">>, #{ <<"targets">> => TargetIds }, State1),
+  State2;
+handle_response({<<"buildTarget/sources">>, _Params}, Response, State) ->
+  handle_sources(apps_paths, fun(Source) -> maps:get(uri, Source, []) end, Response, State);
+handle_response({<<"buildTarget/dependencySources">>, _Params}, Response, State) ->
+  handle_sources(deps_paths, fun(Source) -> Source end, Response, State);
+handle_response(Request, Response, State) ->
+  ?LOG_WARNING("Unhandled response. [request=~p] [response=~p]", [Request, Response]),
+  State.
+
+-spec handle_sources(atom(), fun((any()) -> uri()), map(), state()) -> state().
+handle_sources(ConfigKey, SourceFun, Response, State) ->
+  Result = maps:get(result, Response, #{}),
+  Items = maps:get(items, Result, []),
+  Sources = lists:flatten([ maps:get(sources, Item, []) || Item <- Items ]),
+  Uris = lists:flatten([ SourceFun(Source) || Source <- Sources ]),
+  NewPaths = lists:flatten([ maps:get(path, uri_string:parse(Uri), []) || Uri <- Uris ]),
+  OldPaths = els_config:get(ConfigKey),
+  AllPaths = lists:usort([ els_utils:to_list(P) || P <- OldPaths ++ NewPaths]),
+  els_config:set(ConfigKey, AllPaths),
+  els_indexing:start(),
```

Comment:
Won't this get executed multiple times? Once for apps and once for deps. On top of that, the general provider will also trigger it.

Reply:
Yeah, and it doesn't respect the `initializationOptions` either.

There is some tradeoff here - the BS might take a relatively long time to start up, for example if the project in question has many dependencies to fetch. In that case it might make sense to start indexing right away, as the config file might still contain path entries.

I was hoping the indexing system might be smart enough to detect that a given file hasn't changed since it was last indexed, and that reindexing might be cheap. If not, I think we could make it smarter.

What do you suggest?

Reply:
As a start, I would probably wait for the BS and then follow up with a smarter mechanism if needed. WDYT?